### PR TITLE
made smoler -- and a couple of other things

### PR DESCRIPTION
**Pull request by EvanSchalton**
I should've done this in multiple branches, but here goes:
1. made the `ctx.save_for_backward` return it's params, that made it possible to cut down on lines by not having an extra line just to save the params
2. converted a bunch of the backwards into lambdas, this cut down on the lines for un-marshaling from the save
3. replaced instances of `input` with `_input` since it's a python reserved word

### CODE REVIEW

**Comment by EvanSchalton** [comment]
slooooowly fixing tests, didn't want to leave the PR up with a failing test suite

**Comment by EvanSchalton** [comment]
I probably ran that test_mnist test (and the rest of the test suite) 500 times to find the damn missing asterisks in my code

**Comment by EvanSchalton** [comment]
Probably cut 30-50 lines, I think it's pretty readable -- some of the lambdas might be a bit ugly but I kept that to a minimum

File: tinygrad/ops_ane.py [hunk 1]
```diff
@@ -24,8 +21,8 @@ def compile_relu(ane, sz):
 
 class ReLU(Function):
   @staticmethod
-  def forward(ctx, input):
-    ret = ctx.ane.tensor(input.shape)
-    ctx.ane.run(compile_relu(ctx.ane, input.sz), input, ret)
+  def forward(ctx, _input):
+    ret = ctx.ane.tensor(_input.shape)
+    ctx.ane.run(compile_relu(ctx.ane, _input.sz), _input, ret)
```
Comment (EvanSchalton) [inline tinygrad/ops_ane.py:26]: `input` is a reserved word in python, it's good to avoid overwriting -- even though you probably wont use input

File: tinygrad/ops_cpu.py [hunk 2]
```diff
@@ -35,8 +35,7 @@ def backward(ctx, grad_output):
 class Mul(Function):
   @staticmethod
   def forward(ctx, x, y):
-    ctx.save_for_backward(x, y)
-    return x*y
+    return (lambda x,y: x*y)(*ctx.save_for_backward(x, y))
```
Comment (EvanSchalton) [inline tinygrad/ops_cpu.py:38]: making the `save_for_backwards` return it's params made it possible to do some lambda and decomposition trickery to reduce line counts

**Comment by EvanSchalton** [comment]
I just counted -- I think it's a ~55 line reduction, looking forward to seeing what goodness ghot can squeeze in in another 50 lines
this is going to be great for running models in lambdas without all the overhead

I think the ops could be cleaned up a lot with a config file and a class factory, but I'm not sure if that'd kill readability -- let me know if you're interested in that type of refactor

**Comment by skoshx** [comment]
Im not sure if the save_for_backwards lambda trickery is worth the one line reduction, as it decreases readability...

**Comment by EvanSchalton** [comment]
> Im not sure if the save_for_backwards lambda trickery is worth the one line reduction, as it decreases readability...
@skoshx 
I respect that -- readability is incredibly subjective; I think what's lost is the variable names, but those could easily be replaced with comments.

Readability isn't made worse every where -- there are a lot of places where the code is doing something like this:

```
var = calcVar(params)
ctx.save_for_backwards(var)
return var
```

turning that into:
```
return ctx.save_for_backwards(calcVar(params))[0]
```

doesn't seem to hinder readability, and gets us back two more lines.

the trickier ones -- where readability might be overly hindered are the lambdas; I don't think the sum, pow, etc (basic ops) are bad at all, but I'll admit some of the others might be tricky

IMHO it boils down to a question of 'readable to who?' -- if someone isn't familiar with lambdas this wont be as readable, but IMO -- if someone isn't familiar with lambdas I doubt they're reading a python ML lib.

Would you mind reviewing the PR and indicating which lines you feel like suffer from this change? I'll see if there's a more readable approach to line reduction in those places or revert them if the consensus is that they're not readable

**Comment by kartik4949** [comment]
> Im not sure if the save_for_backwards lambda trickery is worth the one line reduction, as it decreases readability...

Agreed

**Comment by EvanSchalton** [comment]
> > Im not sure if the save_for_backwards lambda trickery is worth the one line reduction, as it decreases readability...
> 
> Agreed

It isn't one line though -- it's 40+
Also, same comment -- care to review the PR and give feedback as to what's not readable?

**Comment by Liamdoult** [comment]
If its just about reducing lines we could just use http://jagt.github.io/python-single-line-convert/. 

Some of the changes are good for example:
``` python
if not isinstance(data, np.ndarray): data = np.array(data, dtype=np.float32)
```

But with equations, it drastically impacts the readability heavily. Here we even have stacked lambdas:
``` python
return (lambda idxs, s: unstack_for_pool(lambda idx: grad_output * (idxs == idx), s, *ctx.kernel_size))(*ctx.saved_tensors)
```

Perhaps these lines can be reduced with a wrapper function that automatically saves the args for backwards and attaches the saved tensors into the backwards function as kwargs?

File: tinygrad/nn.py [hunk 3]
```diff
@@ -3,26 +3,19 @@
...
-
     if self.track_running_stats:
       self.running_mean = (1 - self.momentum) * self.running_mean + self.momentum * batch_mean
       self.running_var = (1 - self.momentum) * self.running_var + self.momentum * batch_var
       self.num_batches_tracked += 1
-
-    if self.training:
-      return self.normalize(x, batch_mean, batch_var)
-
+    if self.training: return self.normalize(x, batch_mean, batch_var)
```
Comment (kartik4949) [inline tinygrad/nn.py:18]: these refactors will not hinder readability, these are good.

File: tinygrad/ops_cpu.py (hunk 2 above)
Comment (kartik4949) [reply tinygrad/ops_cpu.py:38]: this hurts readablity

File: tinygrad/ops_cpu.py [hunk 4]
```diff
@@ -59,9 +58,8 @@ def backward(ctx, grad_output):
 
 class Sum(Function):
   @staticmethod
-  def forward(ctx, input,axis=None):
-    ctx.save_for_backward(input, axis)
-    return np.array([input.sum()]) if axis is None else input.sum(axis=axis)
+  def forward(ctx, _input, axis=None):
+    return (lambda i, a: np.array([i.sum()]) if a is None else _input.sum(axis=a))(*ctx.save_for_backward(_input, axis))
```
Comment (kartik4949) [inline tinygrad/ops_cpu.py:62]: this too

File: tinygrad/ops_cpu.py [hunk 5]
```diff
@@ -109,23 +103,20 @@ def forward(ctx, x, shape):
...
 
 # ************* activation ops *************
 
 class ReLU(Function):
   @staticmethod
-  def forward(ctx, input):
-    ctx.save_for_backward(input)
-    return np.maximum(input, 0)
+  def forward(ctx, _input):
+    return np.maximum(*ctx.save_for_backward(_input), 0)
```
Comment (kartik4949) [inline tinygrad/ops_cpu.py:115]: how can this work? does save_for_backward return anything? may be now

File: tinygrad/ops_cpu.py [hunk 6]
```diff
@@ -141,26 +132,21 @@ def forward(ctx, input):
...
-    ret, = ctx.saved_tensors
-    return grad_output * (ret * (1 - ret))
+    return (lambda r, gout: gout * (r * (1 - r)))(*ctx.saved_tensors, grad_output)
 register('sigmoid', Sigmoid)
 
 class LogSoftmax(Function):
   @staticmethod
-  def forward(ctx, input):
-    softmax = _exp_normalize(input, axis=1)
-    ctx.save_for_backward(softmax)
```
Comment (kartik4949) [inline tinygrad/ops_cpu.py:157]: @EvanSchalton if u want to remove this save_for_backward line, try to bind it (may be make it default to save all inputs) in def forward like a default. (just alternative to what you did and it wont hurt readability either.)
and if someone wants to save different inputs like output of some function e.g Sigmoid forward
they can call save_for_backward again

just a better way to remove that line, but idk if you should remove that or use the above.

File: tinygrad/optim.py [hunk 7]
```diff
@@ -8,23 +8,20 @@ def __init__(self, params):
     self.params = [x for x in params if x.requires_grad == True]
 
   def zero_grad(self):
-    for param in self.params:
-      param.grad = None
+    for param in self.params: param.grad = None
```
Comment (kartik4949) [inline tinygrad/optim.py:11]: this is nice

File: tinygrad/optim.py [hunk 8]
```diff
@@ -8,23 +8,20 @@ def __init__(self, params):
...
 
 class SGD(Optimizer):
   def __init__(self, params, lr=0.001):
     super().__init__(params)
     self.lr = lr
 
   def step(self):
-    for t in self.params:
-      t -= t.grad * self.lr
+    for t in self.params: t -= t.grad * self.lr
```
Comment (kartik4949) [inline tinygrad/optim.py:19]: just a one liner hack idk it works.
 list(map(lambda p :p.assign(p.sub(self.lr*p.grad)) ,self.params))

Comment (marcelbischoff) [reply tinygrad/optim.py:19]: please no `list(map(lambda ...`

Comment (kartik4949) [reply tinygrad/optim.py:19]: > please no `list(map(lambda ...`

yeah just giving alternative ... i also wouldnt prefer oneliners..

File: tinygrad/ops_cpu.py (hunk 2 above)
Comment (EvanSchalton) [reply tinygrad/ops_cpu.py:38]: It kinda just seems like you don't like lambdas, these seem super readable to me

File: tinygrad/ops_cpu.py (hunk 5 above)
Comment (EvanSchalton) [reply tinygrad/ops_cpu.py:115]: Yeah, I updated it to have it return its params - that's how all of my updates work

File: tinygrad/ops_cpu.py (hunk 6 above)
Comment (EvanSchalton) [reply tinygrad/ops_cpu.py:157]: Unfortunately we're most often saving something that's been mutated, that default would probably only be sufficient one or two times

**Comment by geohot** [comment]
If it doesn't improve readability, I don't care about the line count. Many of these changes make the code less readable.

**Comment by EvanSchalton** [comment]
> If it doesn't improve readability, I don't care about the line count. Many of these changes make the code less readable.

Cool; I read the req as reduce line count w/out hurting readability (too much) not as improve readability -- I may take another pass at it later

**Comment by geohot** [comment]
Going to close this, perhaps resubmit in smaller pieces?
